python_modules/libraries/dagster-components/dagster_components/core/component_rendering.py

```diff
@@ -0,0 +1,62 @@
+from typing import Any, Mapping, Optional, Sequence, Type, TypeVar, Union
+
+from jinja2 import Template
+from pydantic import BaseModel
+
+T = TypeVar("T")
+
+REF_BASE = "#/$defs/"
+REF_TEMPLATE = f"{REF_BASE}{{model}}"
+
+
+class Renderer:
+    def __init__(self, context: Mapping[str, Any]):
+        self.context = context
+
+    def render(self, val: str) -> str:
+        return Template(val).render(**self.context)
+
+
+def _should_render(
+    valpath: Sequence[Union[str, int]], json_schema: Optional[Mapping[str, Any]]
+) -> bool:
+    """For a given path to an attribute, read the assoicated metadata in the json_schema to determine if the value should be rendered."""
+    if json_schema is None:
+        return True
+
+    subschema = json_schema
+    for el in valpath:
+        if isinstance(el, str):
+            subschema = subschema["properties"].get(el)
+        else:
+            subschema = subschema["items"]
+
+        if not subschema:
+            return False
+        if "$ref" in subschema:
+            subschema = json_schema["$defs"].get(subschema["$ref"][len(REF_BASE) :])
+        if subschema.get("defer_render") is True:
+            return False
+
+    return True
+
+
+def _render_inner(
+    renderer: Any,
+    val: Any,
+    valpath: Sequence[Union[str, int]],
+    json_schema: Optional[Mapping[str, Any]],
+) -> Any:
+    if not _should_render(valpath, json_schema):
+        return val
+    elif isinstance(val, dict):
+        return {k: _render_inner(renderer, v, [*valpath, k], json_schema) for k, v in val.items()}
+    elif isinstance(val, list):
+        return [_render_inner(renderer, v, [*valpath, i], json_schema) for i, v in enumerate(val)]
+    else:
+        return renderer.render(val)
+
+
+def render(renderer: Renderer, val: T, target_type: Type) -> T:
+    json_schema = target_type.model_json_schema() if issubclass(target_type, BaseModel) else None
```

Comment:
can we make this check an attribute in json schema rather than `BaseModel`?

Reply:
still would like to see this addresses

Reply:
Right now, this feels like the most convenient function signature for the existing use cases, as at some point in the call stack we're going to need to get a json schema that represents a python type, and BaseModel lets us do that easily.

The core logic (`_render_values`) still just operates over json schemas (i.e. the only BaseModel-specific logic here is the call to go from BaseModel->json schema) and so we can always expose a different entry point to that method that is more friendly to that use case in the future